src/bin/sol-fbp-runner/Widget.js

```diff
@@ -0,0 +1,222 @@
+			});
+
+			instance.portsIn.push(port);
+			instance.totalPorts++;
+			port.el.on("port-select", instance.onPortSelect);
+			Widget.prototype.portsCount ++;
+		});
+
+		//reading all out ports
+		payload.ports_out.forEach(function(element, index) {
+			port = new Port(
+			{
+				name:element["name"],
+				value:undefined,
+				type:"out",
+				port_index:element["base_port_idx"],
+				data_type:element["data_type"],
+				required:element["required"],
+				widget:instance,
+				key:instance.key + "_" + Widget.prototype.portsCount
+			});
+			instance.portsOut.push(port);
+			instance.totalPorts++;
+			port.el.on("port-select", instance.onPortSelect);
+			Widget.prototype.portsCount ++;
+		});
+
+		//Auto Adjusting the size of the Widget Container
+		var h = Math.max(this.portsIn.length, this.portsOut.length) * 26;
+		this.widgetPorts.css("height",h);
+
+	}
+
+	Widget.prototype.onPortSelect = function(event) {
+		event.widget.selectPort(event.port, event.value);
+	};
+
+	Widget.prototype.getUID = function() {
+		return this.uid;
+	};
+
+	//UPDATES THE VALUES OF ALL PORTS
+	Widget.prototype.update = function(type, data) {
+
+		//getting the child ports of the widget
+		var port = (type === "deliver") ? this.portsIn[data.port_idx] : this.portsOut[data.port_idx];
+
+		//if the ports have no pair, skip
+		if(port.connections.length < 0) return;
+
+		//found paired ports, update its value
+		var packet = data.packet;
+		var packetType = packet.packet_type;
+
+		if(packetType === "empty"){
+			port.update("empty");
+		}else if(packetType === "int" || packetType === "float"){
+			port.update(packet.payload.value)
+		}else if(packetType === "byte"){
+			port.update(packet.payload);
+		}else if(packetType === "boolean"){
+			port.update(packet.payload.toString());
+		}
+	};
+
+	Widget.prototype.fadeIn = function(delay) {
+		var instance = this;
+		setTimeout(function(){
+			instance.el.fadeIn(300);
+		},delay);
+	};
+
+	Widget.prototype.clearPorts = function() {
+
+		this.portsIn.forEach(function(p, i){
+			if(p.connections.length === 0){
+				p.hidePort();
+			}
+		})
+
+		this.portsOut.forEach(function(p, i){
+			if(p.connections === 0){
+				p.hidePort();
+			}
+		})
+	};
+
+	Widget.prototype.autoSelectPort = function(port) {
+		if(this.selectedPort == null){
+			this.selectPort(port,port.value);
+		}
+	};
+
+	Widget.prototype.selectPort = function(port, value) {
+
+		//dealing with repeated port click
+		if(port === this.selectedPort) {
+			port.unselect();
+			this.unsubscribe(port);
+			this.selectedPort = null;
+			return;
+		}
+
+		//clearing last selected port
+		if(this.selectedPort != null) {
+			this.selectedPort.unselect();
+			this.unsubscribe(port);
+		}
+
+		//selecting new port
+		this.clearDisplay().append(value);
+		this.selectedPort = port;
+		this.selectedPort.select();
+		this.subscribe(port);
+	};
+
+	Widget.prototype.onPortValueChange = function(event) {
+		event.widget.clearDisplay().append(event.value);
```

Comment:
this signature doesn't look right. Why not find the port and specify it here, then call the function on the `Port` so it could update itself.